datafusion/expr/src/type_coercion/functions.rs

```diff
@@ -435,15 +437,15 @@ fn coerced_from<'a>(
         // Note that not all rules in `comparison_coercion` can be reused here.
         // For example, all numeric types can be coerced into Utf8 for comparison,
         // but not for function arguments.
-        _ => comparison_binary_numeric_coercion(type_into, type_from).and_then(
-            |coerced_type| {
+        _ => comparison_binary_numeric_coercion(type_into, type_from)
+            .or_else(|| dictionary_coercion(type_into, type_from, true))
```

Comment:
Should we check the inner data type with `coerced_from` instead of `comparison_coercion` 🤔 ?

Reply:
I couldn't quite tell the difference. What would be the benefit? 

I always get a little confused with the type coercion logic -- that there are different rules for certain operations I think.

Reply:
Hmm, not sure I understand the comment. What you mean "check inner data type with `coerced_from`"?

Reply:
> "check inner data type with coerced_from"

Similar to the current implementation of Dict 
```rust
        // coerced dictionary first
        (cur_type, Dictionary(_, value_type)) | (Dictionary(_, value_type), cur_type)
            if coerced_from(cur_type, value_type).is_some() =>
        {
            Some(type_into.clone())
        }
```
The above once checks if the inner type in Dict is coercible by the `coerced_from` function. 
But `dictionary_coercion` checks the inner type of Dict with `comparison_coercion`.

The `coerced_from` and `comparison_coercion` are slightly different. 
`comparison_coercion` cares about the scenario in comparison, so **loss is allowed**. For example, i64 and u64, we return i64, while we get None in `coerced_from` for casting u64 to i64.

I tried to have one coercion for all but ended up not finding a way to cover all the cases in https://github.com/apache/arrow-datafusion/issues/8302.

I suggest we don't mix the logic for `coerce_from` and `comparison_coercion`. It would be nice to avoid `comparison_binary_numeric_coercion ` too.

Reply:
Ah, I don't notice `coerced_from` is updated to coerce dictionary type. I was working on a branch without update yet. The current update with dictionary coercion looks good now to fix the issue I encountered.

Reply:
But I notice that the current implementation has a small issue.